backend/app/services/paper_ingestion.py

```
@@ -0,0 +1,479 @@
+def _page_number(raw: str | None, current: int | None) -> int | None:
+    if raw and raw.isdigit():
+        return int(raw)
+    return (current + 1) if current is not None else 1
+
+
+def parse_tei(content: bytes, fallback_title: str) -> ParsedPaper:
+    xml = _bounded_gunzip(content)
+    if b"<!DOCTYPE" in xml[:4096].upper():
+        raise IngestionError("unsafe_xml", "XML document types are not accepted.")
+    try:
+        root = ET.fromstring(xml)
```

Ruff (0.15.18)

[error] Using `xml` to parse untrusted data is known to be vulnerable to XML attacks; use `defusedxml` equivalents

(S314)